sw/device/silicon_creator/lib/cert/cert.c

```diff
@@ -52,38 +54,52 @@ rom_error_t cert_x509_asn1_check_serial_number(
   // Tag should be 0x2 (the ASN.1 tag for an INTEGER).
   HARDENED_CHECK_EQ(asn1_tag, kAsn1TagNumberInteger);
-  // Length should be 20 or 21 bytes (depending on if the MSb of the measurement
-  // is 1 since the value is unsigned), as specified in IETF RFC 5280, and
-  // hardcoded in the certificate HJSON specifications.
-  HARDENED_CHECK_GE(asn1_integer_length, kCertX509Asn1SerialNumberSizeInBytes);
+  // Length should be less than or equal to 21 bytes (if the MSb of the
+  // measurement is 1, a zero is added during ASN.1 encoding since the value is
+  // unsigned), as specified in IETF RFC 5280, and hardcoded in the certificate
+  // HJSON specifications.
   HARDENED_CHECK_LE(asn1_integer_length,
                     kCertX509Asn1SerialNumberSizeInBytes + 1);
 
-  // Check the serial number in the certificate matches what was expected.
-  size_t sn_bytes_offset = kCertX509Asn1SerialNumberByteOffset + 2;
+  // If the length is 21 bytes, we skip the first 0 pad byte.
+  size_t sn_bytes_offset = kCertX509Asn1SerialNumberLengthByteOffset + 1;
   if (launder32(asn1_integer_length) ==
       kCertX509Asn1SerialNumberSizeInBytes + 1) {
     HARDENED_CHECK_EQ(asn1_integer_length,
                       kCertX509Asn1SerialNumberSizeInBytes + 1);
     sn_bytes_offset++;
+    asn1_integer_length--;
   }
-  uint32_t curr_sn_word = 0;
-  for (size_t i = 0; i < kCertX509Asn1SerialNumberSizeIn32BitWords; ++i) {
-    curr_sn_word = 0;
-    for (size_t j = 0; j < sizeof(uint32_t); ++j) {
-      curr_sn_word |=
-          (uint32_t)cert_bytes[sn_bytes_offset + (i * sizeof(uint32_t)) + j]
-          << (8 * j);
-    }
-    if (launder32(curr_sn_word) != expected_sn_words[i]) {
-      HARDENED_CHECK_NE(curr_sn_word, expected_sn_words[i]);
+
+  // Extract serial number bytes from the certificate into a 20-byte array as
+  // this is the full size of the serial number integer.
+  //
+  // We fill the end of the array to ensure we maintain any prefix zero padding.
+  //
+  // We make sure to clear out the staging buffer before copying the actual cert
+  // bytes in case this function was previously called on a cert that had a
+  // larger serial number than currently present.
+  memset(actual_serial_number, 0, kCertX509Asn1SerialNumberSizeInBytes);
+  memcpy(&actual_serial_number[kCertX509Asn1SerialNumberSizeInBytes -
+                               asn1_integer_length],
+         &cert_bytes[sn_bytes_offset], asn1_integer_length);
+  /*for (size_t i = sn_bytes_offset + asn1_integer_length - 1,*/
+  /*j = kCertX509Asn1SerialNumberSizeInBytes - 1;*/
+  /*i >= sn_bytes_offset; --i, --j) {*/
+  /*actual_serial_number[j] = cert_bytes[i];*/
+  /*}*/
+
+  // Check the serial number in the certificate matches what was expected.
+  for (size_t i = 0; i < kCertX509Asn1SerialNumberSizeInBytes; ++i) {
+    if (launder32(actual_serial_number[i]) != expected_sn_bytes[i]) {
+      HARDENED_CHECK_NE(actual_serial_number[i], expected_sn_bytes[i]);
```

Comment:
nit: Since this function is pretty long, it may be worth adding `*matches = kHardenedBoolFalse` here.

Feel free to discard this comment on the assumption that this is already covered by tests.

Reply:
It is also covered by unit tests, but good idea, fixed.